common/flatpak-run.c

```diff
@@ -414,8 +421,132 @@ flatpak_run_add_x11_args (FlatpakBwrap         *bwrap,
     }
 }
 
+#ifdef ENABLE_WAYLAND_SECURITY_CONTEXT
+
+static void registry_handle_global (void *data, struct wl_registry *registry,
+                                    uint32_t name, const char *interface,
+                                    uint32_t version)
+{
+  struct wp_security_context_manager_v1 **out = data;
+
+  if (strcmp (interface, wp_security_context_manager_v1_interface.name) == 0) {
+    *out = wl_registry_bind (registry, name,
+                             &wp_security_context_manager_v1_interface, 1);
+  }
+}
+
+static void registry_handle_global_remove (void *data,
+                                           struct wl_registry *registry,
+                                           uint32_t name)
+{
+  /* no-op */
+}
+
+static const struct wl_registry_listener registry_listener = {
+  .global = registry_handle_global,
+  .global_remove = registry_handle_global_remove,
+};
+
+static char *
+create_wl_socket (char *template)
+{
+  g_autofree char *user_runtime_dir = flatpak_get_real_xdg_runtime_dir ();
+  g_autofree char *proxy_socket_dir = g_build_filename (user_runtime_dir, ".flatpak-wl", NULL);
+  g_autofree char *proxy_socket = g_build_filename (proxy_socket_dir, template, NULL);
+  int fd;
+
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, proxy_socket_dir, 0755, NULL, NULL))
+    return NULL;
+
+  fd = g_mkstemp (proxy_socket);
+  if (fd == -1)
+    return NULL;
+
+  close (fd);
+
+  return g_steal_pointer (&proxy_socket);
+}
+
+static gboolean
+flatpak_run_add_wayland_security_context_args (FlatpakBwrap *bwrap,
+                                               const char *app_id)
+{
+  gboolean res = FALSE;
+  struct wl_display *display;
+  struct wl_registry *registry;
+  struct wp_security_context_manager_v1 *security_context_manager = NULL;
+  struct wp_security_context_v1 *security_context;
+  struct sockaddr_un sockaddr = {0};
+  g_autofree char *socket_path = NULL;
+  int listen_fd = -1;
+  int sync_fds[2] = {-1, -1};
+
+  if (!g_getenv ("WAYLAND_DISPLAY"))
+    return FALSE;
+
+  display = wl_display_connect (NULL);
+  if (!display)
+    return FALSE;
+
+  registry = wl_display_get_registry (display);
+  wl_registry_add_listener (registry, &registry_listener,
+                            &security_context_manager);
+  wl_display_roundtrip (display);
+  wl_registry_destroy (registry);
+
+  if (!security_context_manager)
+    goto out;
+
+  socket_path = create_wl_socket ("wayland-XXXXXX");
+  if (!socket_path)
+    goto out;
+
+  listen_fd = socket (AF_UNIX, SOCK_STREAM, 0);
+  if (listen_fd < 0)
+    goto out;
+
+  unlink(socket_path);
+
+  sockaddr.sun_family = AF_UNIX;
+  snprintf (sockaddr.sun_path, sizeof(sockaddr.sun_path), "%s", socket_path);
+  if (bind (listen_fd, (struct sockaddr *) &sockaddr, sizeof (sockaddr)) != 0)
+    goto out;
+
+  if (listen (listen_fd, 0) != 0)
+    goto out;
+
+  if (pipe2 (sync_fds, O_CLOEXEC) < 0)
```

Comment:
I don't see why you can't reuse the existing sync-fd with the dbus proxy. Just needs some shared code in FlatpakBwrap to handle the allocation of it.

Reply:
Done!